src/main/scala/is/hail/expr/types/TableType.scala

```diff
@@ -0,0 +1,24 @@
+package is.hail.expr.types
+
+import is.hail.expr.EvalContext
+import is.hail.expr.ir._
+
+/**
+  * Created by cotton on 1/24/18.
+  */
+case class TableType(rowType: TStruct, key: Array[String], globalType: TStruct) extends BaseType {
+  def rowEC: EvalContext = EvalContext(rowType.fields.map { f => f.name -> f.typ } ++
+      globalType.fields.map { f => f.name -> f.typ }: _*)
+  def fields: Map[String, Type] = Map(rowType.fields.map { f => f.name -> f.typ } ++ globalType.fields.map { f => f.name -> f.typ }: _*)
+
+  def env: Env[Type] = {
+    new Env[Type]()
```

Comment:
I think we should use `Env.empty[Type]` so the constructor can be changed arbitrarily.